tensorflow_datasets/core/features/tensor_feature.py

```diff
@@ -16,12 +16,12 @@
 """Feature connector."""
```

Comment:
Could you run unittests to make sure this is not breaking current generation ? Especially all the ones in `core/features/`

Reply:
In fact, this function has very little correlation with the dataset size. It is directly related to the number of features. 
If we want to be on the safe side, we can limit the cache size.

Reply:
It took some time. Now I do see a lot of errors of unhashable types: TensorShape, list, ...
We can either remove caches or use hashable types...

Reply:
master:
---
11 failed, 2099 passed, 446 skipped, 2617 warnings

PR (with caching removed):
---
========== 19 failed, 2091 passed, 446 skipped, 2617 warnings 


The additional failed tests are expecting failures that are now gone.

Reply:
Our tests are green at master: https://source.cloud.google.com/results/invocations/a220ad4b-b74d-47aa-aadf-c9bda4a658be/targets, so maybe some dependencies are missing from your set-up.

If tests are too long to run, only run the ones in `core/features/` as those should test the feature encoding/decoding.

> It took some time. Now I do see a lot of errors of unhashable types: TensorShape, list, ...
> We can either remove caches or use hashable types...

This means this is gonna break many users which are not using hashable types, so we should revert this change.

Please fix the PR to have tests pass again (at least the ones in `core/features/`).

Reply:
Thanks @Conchylicultor , without the caching this PR has very little value for me. 
I would suggest to standardize some things to make it 
1. more understandable  + fewer surprises
2. more structured 
2. amendable for more optimizations

E.g.,  shapes and bounding boxes must be tuples, elements must be either int64, float, or byte (types supported by the example proto).


If this direction looks good, I can help with the transition.